activerecord/lib/active_record/associations/has_many_through_association.rb

```diff
@@ -38,10 +38,12 @@ def concat_records(records)
       def insert_record(record, validate = true, raise = false)
         ensure_not_nested
 
-        if raise
-          record.save!(:validate => validate)
-        else
-          return unless record.save(:validate => validate)
+        if !record.readonly?
```

Comment:
As this PR is closed, it doesn’t display new commits I added to the branch. So I opened a new PR #24670 with a solution where `save` is actually called from inside AutosaveAssociation.